### ISSUE TITLE
TH2 Calculation might be wrong

### ISSUE THREAD

**Issue by phinexhung** [jyao1/openspdm#148]
TH2 calculation in the following commit seems to be correct..

commit 15133becd1346b62490dcde639bec3621662aed5
Author: Jiewen Yao <jiewen.yao@intel.com>
Date:   Mon Dec 14 10:03:26 2020 +0800

    Use PskHkdfExpandFunction to replace PskHmac.

    This is to hide HandshakeSecret and MasterSecret.

    Signed-off-by: Jiewen Yao <jiewen.yao@intel.com>

But  in the latest commit, it seems to be wrong..

commit 0623e4cab400c69d28b781b14faa18c0166c71fe (HEAD -> master, origin/master, origin/HEAD)
Author: Jiewen Yao <jiewen.yao@intel.com>
Date:   Thu Jan 28 07:52:23 2021 +0800

    Change the link type number according to tcpdump.

    Signed-off-by: Jiewen Yao <jiewen.yao@intel.com>

In the latest commit, it might miss the following item..

 Hash of the specified certificate chain in DER format (i.e., KEY_EXCHANGE Param2)

It only computes MessageA, MessageK, MessageF, but it should be MessageA, MessageCt, MessageK, and MessageF.

**Comment by phinexhung** [comment]
Patch for this can be easily done with...

```
diff --git a/Library/SpdmCommonLib/SpdmCommonLibCryptoServiceSession.c b/Library
/SpdmCommonLib/SpdmCommonLibCryptoServiceSession.c
index 653c7c9..3d0d851 100644
--- a/Library/SpdmCommonLib/SpdmCommonLibCryptoServiceSession.c
+++ b/Library/SpdmCommonLib/SpdmCommonLibCryptoServiceSession.c
@@ -1086,7 +1086,7 @@ SpdmCalculateTH2Hash (

   HashSize = GetSpdmHashSize (SpdmContext->ConnectionInfo.Algorithm.BaseHashAlgo);

-  if (SessionInfo->UsePsk) {
+  if (!SessionInfo->UsePsk) {
     if (IsRequester) {
       Result = SpdmGetPeerCertChainData (SpdmContext, (VOID **)&CertChainData, &CertChainDataSize);
     } else {
```

**Comment by jyao1** [comment]
Thanks. Issue confirmed. My mistake during code move. Do you want to send pull-request?

BTW: How you find the issue?

**Comment by phinexhung** [comment]
Hi @jyao1 ,

I just use openspdm client to test my implementation (actually refer to your old source).

Since tested failed, I just compared all the process of key calculations, and finally found it.

BTW, I also find two extra differences..

1.opaque "DMTF" signature (fixed in the git commit log)
2.add sequence number...

Thanks for your great porting.

**Comment by jyao1** [comment]
thanks. I think there will be 2 incompatible change recently.
please refer to https://github.com/jyao1/openspdm/issues/141 and https://github.com/jyao1/openspdm/issues/143.

Please be aware of that.

**Comment by jyao1** [comment]
Fixed in 04cc8f9befba4f1325cff0c4e701ac0b66bde9ea
